tests/test_aodnfetcher.py

```diff
@@ -354,6 +354,134 @@ def test_custom_jenkins_pattern_to_local_file(self):
+    def test_auth_failure(self):
+        self.fetcher.s3_client.get_object.side_effect = botocore.exceptions.ClientError(
+            {'Error': {'Code': 'AuthorizationHeaderMalformed'}}, 'GetObject')
+        with self.assertRaises(aodnfetcher.fetcherlib.AuthenticationError):
+            _ = self.fetcher.object
+
+    def test_no_builds(self):
+        self.fetcher.s3_client.get_paginator().paginate().result_key_iters.return_value = []
+
+        with self.assertRaises(aodnfetcher.fetcherlib.KeyResolutionError) as cm:
+            _ = self.fetcher.object
+        self.assertEqual(cm.exception.reason_code, 'NO_RESULTS')
+
+    def test_no_matching_keys(self):
+        self.fetcher.s3_client.get_paginator().paginate().result_key_iters.return_value = [
+            [{'Key': 'prefix/3/path3.txt'}]
+        ]
+
+        with self.assertRaises(aodnfetcher.fetcherlib.KeyResolutionError) as cm:
+            _ = self.fetcher.object
+        self.assertEqual(cm.exception.reason_code, 'NO_MATCHING_KEYS')
+
+    def test_custom_filename_pattern(self):
+        url = 's3prefix://bucket/prefix?pattern=^.*\.whl$'
+        fetcher = get_mocked_s3_fetcher(url)
+        fetcher.s3_client.list_objects_v2.__self__ = fetcher.s3_client
+        fetcher.s3_client.list_objects_v2.__name__ = 'list_objects_v2'
+
+        fetcher.s3_client.get_paginator().paginate().result_key_iters.return_value = [
+            [
+                {'Key': 'prefix/1/path1.war', 'LastModified': '2020-07-29 05:41:27+00:00'},
+                {'Key': 'prefix/2/path2.whl', 'LastModified': '2019-07-29 05:41:27+00:00'},
+                {'Key': 'prefix/3/path3.whl', 'LastModified': '2018-07-29 05:41:27+00:00'}
+            ]
+        ]
+
+        self.assertEqual(fetcher.real_url, 's3://bucket/prefix/2/path2.whl')
+
+    def test_custom_filename_pattern_to_local_file(self):
+        url = 's3prefix://bucket/job?pattern=^.*\.whl$&local_file=custom_path.whl'
+        fetcher = get_mocked_s3_fetcher(url)
+        fetcher.s3_client.list_objects_v2.__self__ = fetcher.s3_client
+        fetcher.s3_client.list_objects_v2.__name__ = 'list_objects_v2'
+
+        fetcher.s3_client.get_paginator().paginate().result_key_iters.return_value = [
+            [
+                {'Key': 'prefix/2/path2.whl', 'LastModified': '2019-07-29 05:41:27+00:00'}
+            ]
+        ]
+
+        self.assertEqual(fetcher.real_url, 's3://bucket/prefix/2/path2.whl')
+        self.assertEqual(fetcher.local_file_hint, 'custom_path.whl')
+
+    def test_version_sortmethod(self):
+        url = 's3prefix://bucket/prefix?sortmethod=version'
+        fetcher = get_mocked_s3_fetcher(url)
+        fetcher.s3_client.list_objects_v2.__self__ = fetcher.s3_client
+        fetcher.s3_client.list_objects_v2.__name__ = 'list_objects_v2'
+
+        fetcher.s3_client.get_paginator().paginate().result_key_iters.return_value = [
+            [
+                {'Key': 'prefix/1/version1.war', 'LastModified': '2020-07-29 05:41:27+00:00'},
```

Comment:
Found one more thing during testing... boto3 actually returns LastModified as a `datetime` object, whereas these tests are comparing strings.

In order to match the actual behaviour, it should just be a matter of doing this instead when mocking S3 responses:

```python
# with imports
from datetime import datetime

# in test methods
fetcher.s3_client.get_paginator().paginate().result_key_iters.return_value = [
    [
        {'Key': 'prefix/1/version1.war', 'LastModified': datetime(2020, 7, 29, 5, 41, 27)},
        {'Key': 'prefix/1/version1.war', 'LastModified': datetime(2019, 7, 29, 5, 41, 27)},
        {'Key': 'prefix/1/version1.war', 'LastModified': datetime(2018, 7, 29, 5, 41, 27)}
    ]
]
```